### DISCUSSION TITLE
Native: Style Spec Parity Between Native and Web

### DISCUSSION THREAD

**Discussion by sjg-wdw** [maplibre/maplibre#342] [Native feature proposals]
## Style Spec Parity Between Native and Web

Here's some functionality we'd like to add to MapLibre. This is one proposal of several for Native and if you'd like a little context and our strategy [there's an overview](https://www.wetdogweather.com/blog/maplibre-proposals). This one is purely Native.

At the request of the community, this is an overview of the style spec features supported on the web, but not Native.  The list is based on the [MapLibre Style Spec](https://maplibre.org/maplibre-style-spec/layers/).

### Text and Icon Features

Many of the unsupported features in the Native version of MapLibre are related to icon and text.  Many of them appear to be in the symbol layout logic and would be easier to implement as a group.

| Feature | Comment |
| ------- | ------- |
| icon-overlap | More granular control of icon overlap beyond 'allow' |
| icon-padding (data driven) | Size of area around and icon use for collision detection. |
| text-overlap | More granular control of text overlap beyond 'allow' |
| viewport-glyph | Aligns glyphs to the x-axis of the viewport and places them along a line. |
| text-variable-anchor-offset | Provide a list of offset locations to try before giving up text placement. |

### Expression Support

Expressions are one of the more complex subsystems in MapLibre Native and are the core of the *tangled knot of metaprogramming*.  That said, adding them to the toolkit is easier than figuring out what they're doing or getting them to do it differently.  So, as a group, these should not be too bad.

| Feature | Comment |
| ------- | ------- |
| index-of | Positional return for a string or array of values. |
| slice | Substring of sub-array return.  Classic slice. |
| interpolate-hcl | Interpolate colors using HCL space. |
| interpolate-lab | Interpolate colors using LAB space. |
| number-format | Finer control of number conversion to string. |

### Terrain Related

The terrain changes are actually just adding and tweaking elevation added to DEMs.  This is better covered in another proposal.

### Miscellaneous

The rest of these don't fit into a neat category.  Let's look at each in turn.

The "feature-state" keyword retrieves a data value attached to a given vector or GeoJSON feature.  The toolkit expects those to be set programmatically as opposed to passed in via attributes from the data.

The Native toolkit appears to support feature states, so it's unclear why this isn't implemented.  In theory it should be easy so the fact that it isn't already implemented suggests that it is not.

The "accumulated" keyword relates to GeoJSON clustering and implements a reduce expression within the clusterProperties. This suggests a fair bit of work to get right. 

Lastly, the "fill-extrusion-vertical-gradient" keyword is only implemented on iOS for Native. Presumably this wouldn't be too difficult for Android.

**Comment by HarelM** [comment]
This is a great summary of the parity gap! Thanks!
addProtocol is probably also somewhat related to this topic.
There's also "distance" I think which is missing in web...?

**Reply by sjg-wdw** [reply]
Good to have the "distance" keyword listed somewhere for the web side of things.

Add for addProtocol, we took a stab at a more general solution with the [Data Loading Architecture](https://github.com/maplibre/maplibre/discussions/325).

**Comment by louwers** [comment]
`index-of` and `slice` have been added to the core and to the Android SDK, will be part of the next Android release.

> The Native toolkit appears to support feature states, so it's unclear why this isn't implemented. In theory it should be easy so the fact that it isn't already implemented suggests that it is not.

I didn't even know, that's funny. My guess would be that they simply never gotten around to implementing the Android / iOS SDK.

**Reply by sjg-wdw** [reply]
Great!
